### DISCUSSION TITLE
i found rpc:call from erlang to ergo is not work

### DISCUSSION THREAD

**Discussion by xiaodoudou12** [ergo-services/ergo#114] [Q&A] [answered]
my OTP version is Erlang/OTP 24 [erts-12.1.5] [source] [64-bit] [smp:4:4] [ds:4:4:10] [async-threads:1] [jit].
golang version is go1.18.ergo version is v1.999.211.
i use https://github.com/ergo-services/ergo/blob/v1.999.211/examples/genserver/server.go this to test.
gen_server:cast and gen_server:call is ok,but rpc:call failed.
https://github.com/ergo-services/ergo/blob/v1.999.211/proto/dist/proto.go in this file
```
			case distProtoSPAWN_REQUEST:
				// {29, ReqId, From, GroupLeader, {Module, Function, Arity}, OptList}
				lib.Log("[%s] CONTROL SPAWN_REQUEST [from %s]: %#v", dc.nodename, dc.peername, message.control)
				if message.proxy != nil && message.proxy.session.NodeFlags.EnableRemoteSpawn == false {
					// we didn't allow this feature. proxy session will be closed due to
					// this violation of the contract
					return node.ErrPeerUnsupported
				}
				registerName := ""
				for _, option := range t.Element(6).(etf.List) {
					name, ok := option.(etf.Tuple)
					if !ok || len(name) != 2 {
						return fmt.Errorf("malformed spawn request")
					}
					switch name.Element(1) {
					case etf.Atom("name"):
						registerName = string(name.Element(2).(etf.Atom))
					}
				}

				from := t.Element(3).(etf.Pid)
				ref := t.Element(2).(etf.Ref)

				mfa := t.Element(5).(etf.Tuple)
				module := mfa.Element(1).(etf.Atom)
				function := mfa.Element(2).(etf.Atom)
				var args etf.List
				if str, ok := message.payload.(string); !ok {
					args, _ = message.payload.(etf.List)
				} else {
					// stupid Erlang's strings :). [1,2,3,4,5] sends as a string.
					// args can't be anything but etf.List.
					for i := range []byte(str) {
						args = append(args, str[i])
					}
				}
```

the message value is
```
&dist.distMessage{
control:etf.Tuple{29, etf.Ref{Node:"erl-test_40@10.2.18.22", Creation:0x636ca74d, ID:[5]uint32{0x143cd, 0x843c0001, 0x9a32d1ab, 0x0, 0x0}}, etf.Pid{Node:"erl-test_40@10.2.18.22", ID:0x55, Creation:0x636ca74d}, etf.Pid{Node:"erl-test_40@10.2.18.22", ID:0x47, Creation:0x636ca74d}, etf.Tuple{"erpc", "execute_call", 4}, etf.List{"monitor"}}, payload:etf.List{etf.Ref{Node:"erl-test_40@10.2.18.22", Creation:0x636ca74d, ID:[5]uint32{0x143cb, 0x843c0001, 0x9a32d1ab, 0x0, 0x0}}, "rpc", "request", etf.List{"hello", 3.14}}, proxy:(*dist.proxySession)(nil)}
```
t.Element(6). is  etf.List{"monitor"}, and len(name) != 2, so it failed
i found registerName should be erpc, i don't know is conrrect.

and node/core.go in function func (c *core) RouteSpawnRequest
```
		// spawn new process
		process_opts := processOptions{}
		process_opts.Env = map[gen.EnvKey]interface{}{EnvKeyRemoteSpawn: request.Options}
		process, err_spawn := c.spawn(request.Options.Name, process_opts, b.Behavior, args...)
```
process_opts.Env show init add "ergo:RemoteSpawnRequest" like this
```
process_opts.Env = map[gen.EnvKey]interface{}{EnvKeyRemoteSpawn: request.Options, "ergo:RemoteSpawnRequest":request}
```
because in erlang/rex.go in func (e *erpc) HandleCast will need process.Env("ergo:RemoteSpawnRequest")
```
func (e *erpc) HandleCast(process *gen.ServerProcess, message etf.Term) gen.ServerStatus {
	lib.Log("ERPC [%v]: HandleCast: %#v", process.Self(), message)
	mfa := message.(erpcMFA)
	rsr := process.Env("ergo:RemoteSpawnRequest").(gen.RemoteSpawnRequest)
```
so that,rpc:call from erlang to ergo work ok,but rpc:cast  not work also.

in erlang i use like this
(erl-test_39@10.2.18.22)1> rpc:call('test_39@10.2.18.22', rpc, request, [hello, 3.14]).
[hello,3.14]

**Comment by halturin** [comment, marked as answer]
Its a bug. Created new issue #116

**Comment by halturin** [comment]
anyway, I wouldn't recommend using this feature in production since it has very limited performance (in erlang, too).

**Reply by xiaodoudou12** [reply]
oh, thank you, what about using like this in production
```
gen_server:call({example,'test_client_39@10.2.18.22'}, {echo, 1,2,3}).
gen_server:cast({example,'test_client_39@10.2.18.22'}, {echo, 1,2,3}).
```
gen_server:call and gen_server:cast are ok from erlang to ergo.

**Reply by halturin** [reply]
yeah, its way better to use OTP way.

**Reply by halturin** [reply]
you may also use a regular erlang:send and handle these messages in the HandleInfo callback

**Reply by xiaodoudou12** [reply]
in the examples/erlang/server.go,  erlang to ergo node work ok,but in my other question, ergo to erlang sometimes work ok,and sometimes work failed,
is also better use ServerProcess:Cast and ServerProcess:Call in golang to send message to erlang.but what is `to interface{}` i should set

**Reply by halturin** [reply]
`to` could be:
* etf.Pid 
* etf.Atom("process_name") or just a string "process_name" - for local messaging only
* gen.ProcessID{Name: "process_name", Node: "remote@node.name"} 

and yes, that's correct - to send a cast message, you should use `ServerProcess:Cast(...)`, and to make a call request - `ServerProcess:Call(..)`. You may also use `Process.Send(..)` (or `ServerProcess:Send(..)` which is inherited from Process) to send a regular message in fashion of `erlang:send(..)`

**Reply by xiaodoudou12** [reply]
oh, thank you, i will try it.

**Reply by xiaodoudou12** [reply]
i try it finish.it work like RPCCall and RPCCast, sometimes is ok, sometimes failed.
i use like this
```
to := gen.ProcessID{Name: "ergo_msg_itf", Node: "pathing_pool_39@10.2.18.22"}
args :=  []etf.Term{"ch", 1}
process.Send(to, args)
```
Handshake error: malformed handshake ('a' length of digest) show when failed.

**Reply by halturin** [reply]
I guess I know the problem. I'll take a look at this a bit later.
